File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/TableSource.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.sources

import org.apache.flink.table.api.TableSchema
import org.apache.flink.table.api.types.DataType
import org.apache.flink.table.plan.stats.TableStats
import org.apache.flink.table.util.TableConnectorUtil

/** Defines an external table by providing schema information and used to produce a
  * [[org.apache.flink.api.scala.DataSet]] or [[org.apache.flink.streaming.api.scala.DataStream]].
  * Schema information consists of a data type, field names, and corresponding indices of
  * these names in the data type.
  *
  * To define a TableSource one needs to implement [[TableSource#getReturnType]]. In this case
  * field names and field indices are derived from the returned type.
  *
  * In case if custom field names are required one need to additionally implement
  * the [[DefinedFieldMapping]] trait.
  */
trait TableSource {

  /** Returns the [[DataType]] for the return type of the [[TableSource]]. */
  def getReturnType: DataType

  /** Returns the table schema of the table source */
  def getTableSchema: TableSchema

  /** Describes the table source, it will be used for computing qualified names of
    * [[org.apache.flink.table.plan.schema.FlinkRelOptTable]].
    *
    * Naming Convention that developers MUST follow:
    *
    * "<connector_type_name>-<stream_table_name>"
    *
    * Note: After project push down or filter push down, the explainSource value of new TableSource
    * should be different with the original one.
    *
    *  */
  def explainSource(): String =
    TableConnectorUtil.generateRuntimeName(getClass, getTableSchema.getColumnNames)

  /** Returns the statistics of the table. */
  def getTableStats: TableStats = null

}

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/TableSource.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.sources;

import org.apache.flink.annotation.PublicEvolving;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.table.api.DataTypes;
import org.apache.flink.table.api.TableException;
import org.apache.flink.table.api.TableSchema;
import org.apache.flink.table.types.DataType;
import org.apache.flink.table.utils.TableConnectorUtils;

import static org.apache.flink.table.types.utils.TypeConversions.fromLegacyInfoToDataType;

/**
 * Defines an external table with the schema that is provided by {@link TableSource#getTableSchema}.
 *
 * <p>The data of a {@link TableSource} is produced as a {@code DataSet} in case of a {@code BatchTableSource}
 * or as a {@code DataStream} in case of a {@code StreamTableSource}. The type of ths produced
 * {@code DataSet} or {@code DataStream} is specified by the {@link TableSource#getProducedDataType()} method.
 *
 * <p>By default, the fields of the {@link TableSchema} are implicitly mapped by name to the fields of
 * the produced {@link DataType}. An explicit mapping can be defined by implementing the
 * {@link DefinedFieldMapping} interface.
 *
 * @param <T> The return type of the {@link TableSource}.
 */
@PublicEvolving
public interface TableSource<T> {

	/**
	 * Returns the {@link DataType} for the produced data of the {@link TableSource}.
	 *
	 * @return The data type of the returned {@code DataSet} or {@code DataStream}.
	 */
	default DataType getProducedDataType() {
		final TypeInformation<T> legacyType = getReturnType();
		if (legacyType == null) {
			throw new TableException("Table source does not implement a produced data type.");
		}
		return fromLegacyInfoToDataType(legacyType);
	}

	/**
	 * @deprecated This method will be removed in future versions as it uses the old type system. It
	 *             is recommended to use {@link #getProducedDataType()} instead which uses the new type
	 *             system based on {@link DataTypes}. Please make sure to use either the old or the new type
	 *             system consistently to avoid unintended behavior. See the website documentation
	 *             for more information.
	 */
	@Deprecated
	default TypeInformation<T> getReturnType() {
		return null;
	}

	/**
	 * Returns the schema of the produced table.
	 *
	 * @return The {@link TableSchema} of the produced table.
	 * @deprecated Table schema is a logical description of a table and should not be part of the physical TableSource.
	 *             Define schema when registering a Table either in DDL or in {@code TableEnvironment#connect(...)}.
	 */
	@Deprecated
	TableSchema getTableSchema();

	/**
	 * Describes the table source.
	 *
	 * @return A String explaining the {@link TableSource}.
	 */
	default String explainSource() {
		return TableConnectorUtils.generateRuntimeName(getClass(), getTableSchema().getFieldNames());
	}
}
